src/stan/mcmc/hmc/nuts/base_nuts.hpp

```diff
@@ -60,45 +60,44 @@ namespace stan {
         ps_point z_propose(z_plus);
 
         Eigen::VectorXd p_sharp_plus = this->hamiltonian_.dtau_dp(this->z_);
-        Eigen::VectorXd p_sharp_minus = this->hamiltonian_.dtau_dp(this->z_);
+        Eigen::VectorXd p_sharp_dummy = p_sharp_plus;
+        Eigen::VectorXd p_sharp_minus = p_sharp_plus;
         Eigen::VectorXd rho = this->z_.p;
-        double log_sum_weight = 0;  // log(exp(H0 - H0))
 
+        double log_sum_weight = 0;  // log(exp(H0 - H0))
         double H0 = this->hamiltonian_.H(this->z_);
         int n_leapfrog = 0;
         double sum_metro_prob = 1;  // exp(H0 - H0)
 
         // Build a trajectory until the NUTS criterion is no longer satisfied
         this->depth_ = 0;
-        this->divergent_ = 0;
+        this->divergent_ = false;
 
         while (this->depth_ < this->max_depth_) {
           // Build a new subtree in a random direction
-          Eigen::VectorXd rho_subtree(rho.size());
-          rho_subtree.setZero();
-
+          Eigen::VectorXd rho_subtree = Eigen::VectorXd::Zero(rho.size());
```

Comment:
Won't the original way you had it be more efficient?   I actually don't know if the compiler's clever enough to reuse the right-hand side object given its scope is only the assignment or if this causes an extra malloc and copy.  In which case, the way you were originally doing it would be more efficient.  Probably not worth worrying about.